File: cpp/4_bfs_from_file.cpp
#include <fstream>
#include <iostream>
#include <queue>
#include <string>
#include <utility>
#include <vector>
using namespace std;

int main() {
    // 読み込む迷路ファイル（ここを書き換えると別の迷路を試せます）
    string maze_file = "maze02.txt";

    ifstream fin("../mazes/" + maze_file);

    if (!fin) {
        cerr << "迷路ファイルを開けません: ../mazes/" << maze_file << '\n';
        cerr << "cpp ディレクトリに移動してから実行してください\n";
        return 1;
    }

    int H, W;
    fin >> H >> W;

    int start_row, start_col;
    int goal_row, goal_col;
    fin >> start_row >> start_col;
    fin >> goal_row >> goal_col;

    vector<string> maze(H);
    for (int i = 0; i < H; i++) {
        fin >> maze[i];
    }

    // 各マスまでの距離。-1 は「まだ訪れていない」
    vector<vector<int>> dist(H, vector<int>(W, -1));

    int dr[4] = {-1, 1, 0, 0};
    int dc[4] = {0, 0, -1, 1};

    queue<pair<int, int>> que;

    // スタート地点をキューに入れる
    dist[start_row][start_col] = 0;
    que.push({start_row, start_col});

    while (!que.empty()) {
        auto [row, col] = que.front();
        que.pop();

        for (int i = 0; i < 4; i++) {
            int next_row = row + dr[i];
            int next_col = col + dc[i];

            if (next_row < 0 || next_row >= H ||
                next_col < 0 || next_col >= W) {
                continue;
            }

            if (maze[next_row][next_col] == '#') {
                continue;
            }

            if (dist[next_row][next_col] != -1) {
                continue;
            }

            dist[next_row][next_col] = dist[row][col] + 1;
            que.push({next_row, next_col});
        }
    }

    int answer = dist[goal_row][goal_col];

    cout << maze_file << '\n';

    if (answer == -1) {
        cout << "ゴールには到達できません\n";
    } else {
        cout << "最短距離: " << answer << '\n';
    }

    return 0;
}

File: python/4_bfs_from_file.py
import os
from collections import deque

# 読み込む迷路ファイル（ここを書き換えると別の迷路を試せます）
maze_file = "maze02.txt"

# このプログラムの場所を基準に ../mazes/ を開く（どこから実行しても動くように）
path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "mazes", maze_file)

with open(path, encoding="utf-8") as f:
    H, W = map(int, f.readline().split())
    sy, sx = map(int, f.readline().split())
    gy, gx = map(int, f.readline().split())
    maze = [f.readline().strip() for _ in range(H)]

# 各マスまでの距離。-1 は「まだ訪れていない」
dist = [[-1] * W for _ in range(H)]

dy = [-1, 1, 0, 0]
dx = [0, 0, -1, 1]

que = deque()

# スタート地点をキューに入れる
dist[sy][sx] = 0
que.append((sy, sx))

while que:
    y, x = que.popleft()

    for i in range(4):
        ny = y + dy[i]
        nx = x + dx[i]

        if not (0 <= ny < H and 0 <= nx < W):
            continue

        if maze[ny][nx] == "#":
            continue

        if dist[ny][nx] != -1:
            continue

        dist[ny][nx] = dist[y][x] + 1
        que.append((ny, nx))

answer = dist[gy][gx]

print(maze_file)

if answer == -1:
    print("ゴールには到達できません")
else:
    print("最短距離:", answer)
